kong/conf_loader.lua

```diff
@@ -294,6 +293,40 @@ local function check_and_infer(conf)
     end
   end
 
+  local headers = constants.HEADERS
+  local header_tokens = {
+    [headers.PROXY_LATENCY] = false,
+    [headers.UPSTREAM_LATENCY] = false,
+    [headers.SERVER] = false,
+    [headers.VIA] = false,
+    server_tokens = false,
+    latency_tokens = false,
+  }
+  if conf.headers then
+    for _, token in ipairs(conf.headers) do
+      if token == "off" then
+          break
+
+      elseif header_tokens[token] == nil then
+        errors[#errors+1] = "tokens: invalid entry '" .. tostring(token) .. "'"
+
+      else
+        header_tokens[token] = true
+      end
+    end
+
+    if header_tokens.server_tokens then
+      header_tokens[headers.SERVER] = true
+      header_tokens[headers.VIA] = true
+    end
+
+    if header_tokens.latency_tokens then
+      header_tokens[headers.PROXY_LATENCY] = true
+      header_tokens[headers.UPSTREAM_LATENCY] = true
+    end
+  end
+  conf.headers = header_tokens
```

Comment:
```lua
conf.headers = build_headers_table(conf.headers)
```
?

Reply:
In `check_and_infer`, we should only do validation (the `errors[#errors+1] = ...` part). Below and once we know all values are correct, we can mutate the `conf.headers` table.